qiskit/pulse/filters.py

```diff
@@ -17,48 +17,86 @@
 
-from qiskit.pulse import Schedule
+from qiskit.pulse import Schedule, ScheduleBlock, Instruction
 from qiskit.pulse.channels import Channel
 from qiskit.pulse.schedule import Interval
+from qiskit.pulse.exceptions import PulseError
 
 
 def filter_instructions(
-    sched: Schedule, filters: List[Callable], negate: bool = False, recurse_subroutines: bool = True
-) -> Schedule:
+    sched: Union[Schedule, ScheduleBlock],
+    filters: List[Callable],
+    negate: bool = False,
+    recurse_subroutines: bool = True,
+) -> Union[Schedule, ScheduleBlock]:
     """A filtering function that takes a schedule and returns a schedule consisting of
     filtered instructions.
 
     Args:
-        sched: A pulse schedule to be filtered.
+        sched: A pulse schedule or schedule_block to be filtered.
         filters: List of callback functions that take an instruction and return boolean.
         negate: Set `True` to accept an instruction if a filter function returns `False`.
             Otherwise the instruction is accepted when the filter function returns `False`.
         recurse_subroutines: Set `True` to individually filter instructions inside of a subroutine
             defined by the :py:class:`~qiskit.pulse.instructions.Call` instruction.
 
     Returns:
-        Filtered pulse schedule.
+        Filtered pulse schedule or schedule_block.
     """
     from qiskit.pulse.transforms import flatten, inline_subroutines
 
-    target_sched = flatten(sched)
-    if recurse_subroutines:
-        target_sched = inline_subroutines(target_sched)
+    if isinstance(sched, Schedule):
+        target_sched = flatten(sched)
+        if recurse_subroutines:
+            target_sched = inline_subroutines(target_sched)
 
-    time_inst_tuples = np.array(target_sched.instructions)
+        time_inst_tuples = np.array(target_sched.instructions)
 
-    valid_insts = np.ones(len(time_inst_tuples), dtype=bool)
-    for filt in filters:
-        valid_insts = np.logical_and(valid_insts, np.array(list(map(filt, time_inst_tuples))))
+        valid_insts = np.ones(len(time_inst_tuples), dtype=bool)
+        for filt in filters:
+            valid_insts = np.logical_and(valid_insts, np.array(list(map(filt, time_inst_tuples))))
 
-    if negate and len(filters) > 0:
-        valid_insts = ~valid_insts
+        if negate and len(filters) > 0:
+            valid_insts = ~valid_insts
 
-    filter_schedule = Schedule.initialize_from(sched)
-    for time, inst in time_inst_tuples[valid_insts]:
-        filter_schedule.insert(time, inst, inplace=True)
+        filter_schedule = Schedule.initialize_from(sched)
+        for time, inst in time_inst_tuples[valid_insts]:
+            filter_schedule.insert(time, inst, inplace=True)
 
-    return filter_schedule
+        return filter_schedule
+
+    elif isinstance(sched, ScheduleBlock):
+        target_sched_blk = sched
+        if recurse_subroutines:
+            target_sched_blk = inline_subroutines(target_sched_blk)
```

Comment:
The [implementation](https://github.com/Qiskit/qiskit-terra/blob/3ebef17322c7714aedaa71967dee74fd455a83a0/qiskit/pulse/transforms/canonicalization.py#L206-L209) of this function intentionally preserves nested schedule blocks because they may have different alignment context (we cannot expose blocks to outer context without breaking alignment). This may cause unexpected behavior, since block to be filtered could be nested ScheduleBlock rather than instructions. I think you need a function that recursively returns all blocks regardless of alignment. I think filtered instructions doesn't need to preserve the timing information as I commented on the issue.